File: 1976.py
import sys
input = sys.stdin.readline

def findParent(parent, node):
  if parent[node] == node:
    return node
  else:
    return findParent(parent, parent[node])

def unionParent(parent, node1, node2):
  node1 = findParent(parent, node1)
  node2 = findParent(parent, node2)
  if node1 < node2:
    parent[node2] = node1
  else:
    parent[node1] = node2

N = int(input())
M = int(input())
parent = [i for i in range(N+1)]
graph = []
for _ in range(N):
  graph.append(list(map(int, input().split())))

for i in range(N):
  for j in range(N):
    if graph[i][j] == 1 and parent[i+1] != parent[j+1]:
      unionParent(parent, i+1, j+1)

route = list(map(int, input().split()))

def isPossibleRoute(parent, route):
  target = findParent(parent, parent[route[0]])
  for i in range(M):
    if findParent(parent, parent[route[i]]) != target:
      return "NO"
  return "YES"

print(isPossibleRoute(parent, route))

File: 1976.java
import java.io.*;

public class Main {
    
    public static int[] parent = new int[201];
    
    public static void main(String args[]) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        BufferedWriter bw = new BufferedWriter(new OutputStreamWriter(System.out));
        
        int N = Integer.parseInt(br.readLine());
        int M = Integer.parseInt(br.readLine());
        boolean isConnected = true;
        
        for(int i=1; i<= N; i++) {
            parent[i] = i;
        }
        
        for(int i=1; i<=N; i++) {
            String[] connection = br.readLine().split(" ");
            for(int j=0; j<N; j++) {
                if(connection[j].equals("1")) {
                    unionParent(i, j+1);
                }
            }
        }
        
        String[] plan = br.readLine().split(" ");
        int targetParent = findParent(Integer.parseInt(plan[0]));
        for(int i=1; i<M; i++) {
            if(findParent(Integer.parseInt(plan[i])) != targetParent) {
                bw.write("NO");
                isConnected = false;
                break;
            }
        }
        
        if(isConnected) {
            bw.write("YES");
        }
        
        bw.flush();
        bw.close();
    }
    
    static int findParent(int node) {
        if(parent[node] == node) {
            return node;
        }
        return findParent(parent[node]);
    }
    
    static void unionParent(int node1, int node2) {
        node1 = findParent(node1);
        node2 = findParent(node2);
        
        if(node1 < node2) {
            parent[node2] = node1;
        }
        else {
            parent[node1] = node2;
        }
    }
}
